src/gui/properties/peerlistwidget.cpp

```diff
@@ -262,44 +262,53 @@ void PeerListWidget::updatePeerCountryResolutionState()
+            return;
+
+        action->setEnabled(false);
+        action->setToolTip(tooltip);
+    };
+
+    auto *menu = new QMenu(this);
     menu->setAttribute(Qt::WA_DeleteOnClose);
+    menu->setToolTipsVisible(true);
 
     // Add Peer Action
-    // Do not allow user to add peers in a private torrent
-    if (!torrent->isQueued() && !torrent->isChecking() && !torrent->isPrivate())
+    QAction *addNewPeer = menu->addAction(UIThemeManager::instance()->getIcon("user-group-new"), tr("Add a new peer...")
+        , this, [this, torrent]()
     {
-        menu->addAction(UIThemeManager::instance()->getIcon("user-group-new"), tr("Add a new peer...")
-            , this, [this, torrent]()
+        const QVector<BitTorrent::PeerAddress> peersList = PeersAdditionDialog::askForPeers(this);
+        const int peerCount = std::count_if(peersList.cbegin(), peersList.cend(), [torrent](const BitTorrent::PeerAddress &peer)
         {
-            const QVector<BitTorrent::PeerAddress> peersList = PeersAdditionDialog::askForPeers(this);
-            const int peerCount = std::count_if(peersList.cbegin(), peersList.cend(), [torrent](const BitTorrent::PeerAddress &peer)
-            {
-                return torrent->connectPeer(peer);
-            });
-            if (peerCount < peersList.length())
-                QMessageBox::information(this, tr("Adding peers"), tr("Some peers cannot be added. Check the Log for details."));
-            else if (peerCount > 0)
-                QMessageBox::information(this, tr("Adding peers"), tr("Peers are added to this torrent."));
+            return torrent->connectPeer(peer);
         });
-    }
+        if (peerCount < peersList.length())
+            QMessageBox::information(this, tr("Adding peers"), tr("Some peers cannot be added. Check the Log for details."));
+        else if (peerCount > 0)
+            QMessageBox::information(this, tr("Adding peers"), tr("Peers are added to this torrent."));
+    });
+    adjustAction(addNewPeer, torrent->isChecking(), tr("Cannot add peers when torrent is checking"));
+    adjustAction(addNewPeer, torrent->isQueued(), tr("Cannot add peers when torrent is queued"));
+    adjustAction(addNewPeer, torrent->isPrivate(), tr("Cannot add peers to a private torrent"));
 
-    if (!selectionModel()->selectedRows().isEmpty())
-    {
-        menu->addAction(UIThemeManager::instance()->getIcon("edit-copy"), tr("Copy IP:port")
-            , this, &PeerListWidget::copySelectedPeers);
-        menu->addSeparator();
-        menu->addAction(UIThemeManager::instance()->getIcon("user-group-delete"), tr("Ban peer permanently")
-            , this, &PeerListWidget::banSelectedPeers);
-    }
+    const bool isSelectionEmpty = selectionModel()->selectedRows().isEmpty();
 
-    if (menu->isEmpty())
-        delete menu;
-    else
-        menu->popup(QCursor::pos());
+    QAction *copyPeers = menu->addAction(UIThemeManager::instance()->getIcon("edit-copy"), tr("Copy IP:port")
```

Comment:
Also you can place all action creations together and all adjustAction call together as well.

Reply:
OK, done.